src/book/book.rs

```diff
@@ -0,0 +1,365 @@
+/// `SUMMARY.md` give the chapter locations relative to it.
+fn load_book_from_disk<P: AsRef<Path>>(summary: &Summary, src_dir: P) -> Result<Book> {
+    debug!("[*] Loading the book from disk");
+    let src_dir = src_dir.as_ref();
+
+    let prefix = summary.prefix_chapters.iter();
+    let numbered = summary.numbered_chapters.iter();
+    let suffix = summary.suffix_chapters.iter();
+
+    let summary_items = prefix.chain(numbered).chain(suffix);
+
+    let chapters = summary_items
+        .map(|i| load_summary_item(i, src_dir))
+        .collect::<Result<_>>()
+        .chain_err(|| "Couldn't load chapters from disk")?;
+
+    Ok(Book { sections: chapters })
+}
+
+fn load_summary_item<P: AsRef<Path>>(item: &SummaryItem, src_dir: P) -> Result<BookItem> {
+    match *item {
+        SummaryItem::Separator => Ok(BookItem::Separator),
+        SummaryItem::Link(ref link) => load_chapter(link, src_dir).map(|c| BookItem::Chapter(c)),
+    }
+}
+
+fn load_chapter<P: AsRef<Path>>(link: &Link, src_dir: P) -> Result<Chapter> {
+    debug!("[*] Loading {} ({})", link.name, link.location.display());
+    let src_dir = src_dir.as_ref();
+
+    let location = if link.location.is_absolute() {
+        link.location.clone()
+    } else {
+        src_dir.join(&link.location)
+    };
+
+    let mut f = File::open(&location).chain_err(|| {
+        format!("Chapter file not found, {}", link.location.display())
+    })?;
+
+    let mut content = String::new();
+    f.read_to_string(&mut content)?;
+
+    let stripped = location.strip_prefix(&src_dir).expect("Chapters are always inside a book");
+
+    let mut ch = Chapter::new(&link.name, content, stripped);
+    ch.number = link.number.clone();
+
+    let sub_items = link.nested_items
+        .iter()
+        .map(|i| load_summary_item(i, src_dir))
+        .collect::<Result<Vec<_>>>()?;
+
+    ch.sub_items = sub_items;
+
+    Ok(ch)
+}
+
+/// A depth-first iterator over the items in a book.
+///
+/// # Note
+///
+/// This struct shouldn't be created directly, instead prefer the
+/// [`Book::iter()`] method.
+///
+/// [`Book::iter()`]: struct.Book.html#method.iter
+pub struct BookItems<'a> {
+    items: VecDeque<&'a BookItem>,
+}
+
+impl<'a> Iterator for BookItems<'a> {
+    type Item = &'a BookItem;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let item = self.items.pop_front();
+
+        if let Some(&BookItem::Chapter(ref ch)) = item {
+            // if we wanted a breadth-first iterator we'd `extend()` here
+            for sub_item in ch.sub_items.iter().rev() {
+                self.items.push_front(sub_item);
+            }
+        }
+
+        item
+    }
+}
+
+impl Display for Chapter {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        if let Some(ref section_number) = self.number {
+            write!(f, "{} ", section_number)?;
+        }
+
+        write!(f, "{}", self.name)
+    }
+}
+
+
+#[cfg(test)]
```

Comment:
In another project, I have successfully placed tests in their own files. In this case we would have a file `src/book/tests.rs` containing all the tests. This clearly separates the tests from the code and reduces the size of the files, especially if we have a lot of tests.

Reply:
True, but it also splits them up from the module they're testing so then you'd have issues where you can't directly test the private methods, wouldn't it?

From what I've seen it's common practice to put unit tests in the same file as the things they're testing instead of moving them out into their own files, so that's why I originally put the tests where they are.

Reply:
>you'd have issues where you can't directly test the private methods

Yeah I don't remember the rules exactly. Sub-modules can access private methods but I am not sure about sibling modules. 

>From what I've seen it's common practice to put unit tests in the same file as the things they're testing

Certainly, but when the number of lines of tests largely exceed the number of lines of code, I am not necessarily fond of this convention. But this is not a blocker for the PR, I don't mind if we merge it like this and see if we can improve the situation in the future if it becomes unwieldy.

Reply:
we can always decide to split the tests along the lines of public interface / implementation details. Impl detail tests could live next to the code under test and the "smoke tests" of the public interface could be moved into a separate file.